### PR TITLE
feat: add DIP-14 256-bit child key derivation (Derive256)

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented

[DIP-14](https://github.com/dashpay/dips/blob/master/dip-0014.md) extends BIP32 child key derivation to 256-bit child indexes with an explicit hardened flag. It underpins the DIP-13 platform key hierarchy and DIP-15 friendship-derived address spaces, where child indexes are 256-bit identity hashes so that the derivation path between two identities cannot be ground out by a third party (a 31-bit index space would be brute-forceable by cycling identities).

Dash Core currently has no DIP-14 primitive. This PR adds it at the `CKey`/`CPubKey` level, extracted from the Platform GUI work in PastaPastaPasta#49 as the first wallet-side prerequisite.

## What was done?

- `DIP14Hash` in `hash.{h,cpp}`: the HMAC-SHA512 with a 256-bit big-endian (`ser256`) child index, alongside `BIP32Hash`.
- `CKey::Derive256`: private child derivation with a 256-bit index and explicit hardened flag.
- `CPubKey::Derive256`: the public (non-hardened only) counterpart. Non-hardened 256-bit public derivation matches private derivation, which is what lets a counterparty derive addresses from an exported xpub without seeing private keys.

**Compatibility-mode semantics** (worth calling out, since the DIP-14 pseudocode is loosely worded here): indexes below 2^32 are interpreted as raw BIP32 index space, per the DIP's Compatibility section ("if a child key's index is less than 2^32 … will match the derivation outlined in BIP32") and its published test vectors. Concretely, the hardened flag is folded into the high bit on the private side, and the public side rejects indexes with the high bit set exactly as BIP32 does. Vector 2 (`m/9'/5'/15'/0'/…`) only reproduces under this interpretation.

## How Has This Been Tested?

New unit suite `dip14_tests`:
- the four DIP-14 spec test vectors from dashpay/dips dip-0014.md, including the mixed 32-bit/256-bit hardened path;
- BIP32 compatibility of the sub-2^32 fallback (private hardened/non-hardened and public, checked against `Derive`);
- public/private derivation consistency over a 256-bit two-step path, and rejection of hardened indexes on the public side.

`dip14_tests`, `key_tests`, `bip32_tests`, and `hash_tests` pass locally (macOS, clang). `lint-whitespace`, `lint-include-guards`, `lint-includes` clean; the new Dash-specific test file is listed in `test/util/data/non-backported.txt`.

## Breaking Changes

None. Additive API only; no existing derivation path changes behavior.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit 77453314)

File: src/key.h [hunk 1]
```diff
@@ -157,6 +157,11 @@ class CKey
     //! Derive BIP32 child key.
     [[nodiscard]] bool Derive(CKey& keyChild, ChainCode &ccChild, unsigned int nChild, const ChainCode& cc) const;
 
+    //! Derive DIP-14 child key with a 256-bit index (32 bytes, big-endian).
+    //! Indexes below 2^32 fall back to BIP32 derivation for compatibility
+    //! (the hardened flag is then folded into the 32-bit index).
+    [[nodiscard]] bool Derive256(CKey& keyChild, ChainCode& ccChild, Span<const unsigned char> nChild, bool hardened, const ChainCode& cc) const;
```
Comment (thepastaclaw) [inline src/key.h:163]: **💬 Nitpick: Span<const unsigned char> for a fixed-32-byte index relies on a runtime assert instead of compile-time size safety**

Derive256 takes Span<const unsigned char> nChild and only validates its length with assert(nChild.size() == 32) at the top of the function body (src/key.cpp:311, src/pubkey.cpp:263). A std::array<unsigned char, 32> (or a dedicated 32-byte index type) would make the 32-byte contract enforced by the type system rather than a runtime assertion, removing any reliance on assert being compiled in for safety against a mis-sized Span reaching nChild.begin() + 28 / nChild.data() + 28. This is consistent with existing Span usage elsewhere in the file, so it's a minor style preference rather than a defect.

<sub>source: ['claude']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Codex + Sonnet

This PR adds a clean, additive DIP-14 256-bit child key derivation primitive (CKey::Derive256 / CPubKey::Derive256 / DIP14Hash) that correctly falls back to BIP32 semantics for sub-2^32 indexes and is verified against published test vectors. The implementation, build registration, and non-backported tracking are all correct; only two minor test-robustness/style suggestions were found, neither blocking.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `claude-sonnet-5` — final-verifier
- Sonnet reviewers: `claude-sonnet-5` — general (completed), `claude-sonnet-5` — dash-core-commit-history (failed), `claude-sonnet-5` — dash-core-commit-history (completed)

🟡 1 suggestion(s) | 💬 1 nitpick(s)

_1 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/test/fuzz/key.cpp`:
- [SUGGESTION] src/test/fuzz/key.cpp:88-97: New Derive256 primitive is not exercised by the existing key/pubkey fuzz harnesses
  src/test/fuzz/key.cpp already fuzzes CKey::Derive and CPubKey::Derive with randomized ChainCode/key material (lines 91-97 and around 240-244), which has historically caught edge cases in secp256k1 tweak-add failure paths and boundary values. Derive256 is new parsing/arithmetic surface — the 28-zero-byte BIP32-compatibility check, ReadBE32 boundary handling, and hardened-bit folding — that would benefit from the same fuzzing treatment, especially since Span<const unsigned char> callers could pass adversarial 32-byte indexes (all-zero, all-0xFF, boundary values around 2^32) not covered by the four fixed spec vectors in dip14_tests.cpp. This is confirmed: grep shows Derive256 is only exercised in src/test/dip14_tests.cpp, never in the fuzz targets. Not a defect in the current code, but a natural follow-up for parity with the existing harness.
```
</details>

File: src/pubkey.cpp [hunk 2]
```diff
@@ -258,6 +258,34 @@ bool CPubKey::Derive(CPubKey& pubkeyChild, ChainCode &ccChild, unsigned int nChi
     return true;
 }
 
+bool CPubKey::Derive256(CPubKey& pubkeyChild, ChainCode& ccChild, Span<const unsigned char> nChild, const ChainCode& cc) const {
+    assert(IsValid());
+    assert(nChild.size() == 32);
+    assert(size() == COMPRESSED_SIZE);
+    // DIP-14 compatibility mode: indexes below 2^32 derive exactly as BIP32.
+    // A hardened (high bit set) 32-bit index cannot be derived from a pubkey.
+    if (std::all_of(nChild.begin(), nChild.begin() + 28, [](unsigned char c) { return c == 0; })) {
```
Comment (kwvg) [inline src/pubkey.cpp:267]: Am not confident about keeping a compatibility mode, since `Derive256` now has an implicit case purely based on input and if we have behaviour assuming that `Derive256` will always emit a DIP-14 derivation, silently switching to BIP-32 derivation might not be ideal, would be preferable to fail entirely.

File: src/hash.cpp [hunk 3]
```diff
@@ -75,6 +75,11 @@ void BIP32Hash(const ChainCode &chainCode, unsigned int nChild, unsigned char he
     CHMAC_SHA512(chainCode.begin(), chainCode.size()).Write(&header, 1).Write(data, 32).Write(num, 4).Finalize(output);
 }
 
+void DIP14Hash(const ChainCode& chainCode, const unsigned char nChild[32], unsigned char header, const unsigned char data[32], unsigned char output[64])
```
Comment (kwvg) [inline src/hash.cpp:78]: We could probably create a type for children that can accept 32-bit or 256-bit integers and encodes them appropriately embedding the BE semantics as part of the wire definition. Would also apply to the `WriteBE32()` in `BIP32Hash()`.

File: src/key.h (hunk 1 above)
Comment (PastaPastaPasta) [reply src/key.h:163]: Addressed in f16cbb97d53b. I kept the byte-oriented Span API so callers are not unnecessarily constrained to one container type, but replaced the assertions with unconditional 32-byte validation in both private and public derivation paths. New 31-byte and 33-byte regression cases verify that release builds reject malformed input before any offset arithmetic.

---
🤖 Posted autonomously by Codex on behalf of pasta.

File: src/pubkey.cpp (hunk 2 above)
Comment (PastaPastaPasta) [reply src/pubkey.cpp:267]: Keeping compatibility mode is intentional here. DIP-14 defines the below-2^32 compatibility behavior, and this PR's vectors explicitly lock that contract to BIP-32. In the DashPay PR train, higher-level callers do not expose this as an untyped choice: wallet::platformkeys::PathElement distinguishes uint32_t from the 32-byte identity path element and routes them to Derive and Derive256 respectively. Removing compatibility in this low-level primitive would diverge from the stated DIP-14 contract without improving the later caller boundary.

---
🤖 Posted autonomously by Codex on behalf of pasta.

File: src/hash.cpp (hunk 3 above)
Comment (PastaPastaPasta) [reply src/hash.cpp:78]: That typed distinction belongs at the wallet path-policy layer, and the later DashPay train already supplies it: wallet::platformkeys::PathElement is a variant of uint32_t and std::array<uint8_t, 32>, routing each case to Derive or Derive256. Pulling that policy type into BIP32Hash/CKey in this foundational PR would couple the low-level crypto API to a later wallet abstraction and widen the dependency PR. I am therefore leaving this primitive byte-oriented and keeping the type-safe selection in the downstream layer.

---
🤖 Posted autonomously by Codex on behalf of pasta.

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Codex/Sol only (Phase 2 disabled)

The current head correctly adds DIP-14 derivation and now rejects malformed child-index spans before performing offset arithmetic; no correctness or security defects were confirmed. Two non-blocking improvements remain: extend the existing key fuzz harness to cover Derive256 and squash the immediate input-validation correction into the feature commit.
Source: reviewers gpt-5.6-sol (general) and gpt-5.6-sol (dash-core-commit-history); final verifier gpt-5.6-sol. openclaw-agent/cliproxy/gpt-5.6-sol is orchestration-only and not reviewer evidence.

Validated zero-blocker Codex/Sol precheck evidence was promoted to final because Phase 2 (Sonnet/Opus) is temporarily disabled. This is Codex/Sol-only final validation, not Codex + Sonnet/Opus coverage.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet/Opus: **not run (Phase 2 disabled — temporary Codex/Sol-only final)**
- Secondary pass: disabled (`temporary_phase2_sonnet_disable`)

🟡 2 suggestion(s)

_2 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/test/fuzz/key.cpp`:
- [SUGGESTION] src/test/fuzz/key.cpp:93: New Derive256 primitive is not exercised by the existing key/pubkey fuzz harnesses
  The key fuzz target exercises CKey::Derive here and CPubKey::Derive around line 242, but it never invokes the newly added Derive256 methods. Add private and public derivations with randomized 32-byte indexes so fuzzing reaches the BIP32 compatibility check, the 32-bit boundary, hardened-bit handling, general 256-bit indexes, and secp256k1 tweak-add failures. The fixed unit tests validate published vectors and malformed lengths, but they do not provide the broad valid-index coverage already given to the existing Derive paths.

In `<commit:f16cbb9>`:
- [SUGGESTION] <commit:f16cbb9>:1: Squash the child-index size fix into the feature commit
  Commit f16cbb97d53 changes assertions in the Derive256 implementations introduced by the immediately preceding commit 8a6299d1ca8 into unconditional input validation and adds the corresponding regression tests. Because the primitive has not shipped between these commits, squash the corrective commit into the feature commit so the API and its complete malformed-input behavior form one atomic change for history, blame, and bisect.
```
</details>

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Codex/Sol only (Phase 2 disabled)

At exact head 7745331468e0abdde9c83e3e9c3d981c37f9623a, the PR cleanly adds the DIP-14 hash and private/public derivation primitives with published-vector, compatibility, malformed-input, consistency, and fuzz coverage. Both prior suggestions are fixed at the current head, and no in-scope correctness, security, build-registration, or test-coverage findings remain.
Source: reviewers gpt-5.6-sol (general) and gpt-5.6-sol (dash-core-commit-history); final verifier gpt-5.6-sol. openclaw-agent/cliproxy/gpt-5.6-sol is orchestration-only and not reviewer evidence.

Validated zero-blocker Codex/Sol precheck evidence was promoted to final because Phase 2 (Sonnet/Opus) is temporarily disabled. This is Codex/Sol-only final validation, not Codex + Sonnet/Opus coverage.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet/Opus: **not run (Phase 2 disabled — temporary Codex/Sol-only final)**
- Secondary pass: disabled (`temporary_phase2_sonnet_disable`)
